module/zfs/dsl_dataset.c

```diff
@@ -1644,23 +1646,59 @@ dsl_dataset_snapshot_check(void *arg, dmu_tx_t *tx)
 			if (error == 0)
 				(void) strlcpy(dsname, name, atp - name + 1);
 		}
-		if (error == 0)
+		if (error == 0) {
 			error = dsl_dataset_hold(dp, dsname, FTAG, &ds);
+			dsheld = (error == 0);
+		}
 		if (error == 0) {
 			/* passing 0/NULL skips dsl_fs_ss_limit_check */
 			error = dsl_dataset_snapshot_check_impl(ds,
 			    atp + 1, tx, B_FALSE, 0, NULL);
-			dsl_dataset_rele(ds, FTAG);
 		}
 
 		if (error != 0) {
 			if (ddsa->ddsa_errors != NULL) {
 				fnvlist_add_int32(ddsa->ddsa_errors,
 				    name, error);
 			}
+			if (dsheld)
+				dsl_dataset_rele(ds, FTAG);
 			rv = error;
+		} else {
+			list_insert_tail(&datasets, ds);
 		}
 	}
+	if (rv != 0) {
+		while ((check_ds = list_remove_head(&datasets)) != NULL) {
+			dsl_dataset_rele(check_ds, FTAG);
+		}
+		list_destroy(&datasets);
+		return (rv);
+	}
+
+	/* Now reserve space since we passed all the checks. */
+	while ((check_ds = list_remove_head(&datasets)) != NULL) {
+		int error = 0;
+
+		/*
+		 * This accumulates dsl_dir_t dd_space_to_write which may
+		 * need to be undone if we get ENOSPC while reserving.
+		 */
+		error = dsl_dataset_snapshot_reserve_space(check_ds, tx);
+		dsl_dataset_rele(check_ds, FTAG);
+		if (error != 0)
+			rv = error;
+	}
+	list_destroy(&datasets);
+	if (rv == 0)
+		return (rv);
+
+	ASSERT3U(rv, ==, ENOSPC);
+	dsl_dir_t *dd;
+	/* Undo the reserved space since no snapshot will be taken. */
+	while ((dd = txg_list_remove(&dp->dp_dirty_dirs, tx->tx_txg)) != NULL) {
```

Comment:
Suggestion by @tcaputi - don't remove all the dp_dirty_dirs, just the ones that we dirtied in the reserve space loop.

Reply:
Constraining the undoing of reserved space to just the ds->ds_dir of the ones we dirtied didn't work.  There were panics in the test log.  A few thoughts... 1) this is a global space check in sync context, so all the snaps will either happen or they won't.  2) dsl_dir_willuse_space dirties the ds->ds_dir AND all its parents and 3)  dp_dirty_dirs must be empty as is ASSERTed in spa_sync.